cpp/src/arrow/array/array_dict.h

```diff
@@ -20,17 +20,23 @@
 #include <cstdint>
 #include <memory>
 
+#include "arrow/array.h"
 #include "arrow/array/array_base.h"
+#include "arrow/array/array_primitive.h"
 #include "arrow/array/data.h"
 #include "arrow/result.h"
 #include "arrow/scalar.h"
 #include "arrow/status.h"
 #include "arrow/type.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/logging.h"
 #include "arrow/util/macros.h"
 #include "arrow/util/visibility.h"
 
 namespace arrow {
 
+using ViewType = std::variant<std::string_view, int32_t>;
```

Comment:
If we use a variant then it's going to have so many variants that it's essentially the `Scalar` class